### PR TITLE
Add Schema examples annotation

### PR DESCRIPTION
**Pull request by gcheadle-vmware**
- [x] include annotation lines in mismatched type error
- [x] collect mismatched type error all provided examples and return together

### CODE REVIEW

File: pkg/schema/annotations.go
Comment (cari-lynn) [inline pkg/schema/annotations.go]: I am curious about this line. It looks like we are concatenating all the `description` annotations values together here into one string. But I was under the impression that only the first annotation will be used if there are multiple of the same type. Whats the story here?

**Review by cari-lynn — APPROVED** [review]
Great job covering all the possible test cases. I'm really happy with the code added here, and think the `description` struct is a good way to group these annotations. 

One piece for discussion:

There is two different syntax for adding an example WITHOUT a description.

`@schema/examples (‘value’,)`

 `@schema/examples (“”, “value”)`
Since if the description is `“”` it will be ignored.

I do like the readability of the second one, I can't imagine a use case where the description needs to be `""`. But I do wonder if it is a good idea to remove the string just because it is zero value.. 🤔

File: pkg/schema/annotations.go
Comment (gcheadle-vmware) [reply pkg/schema/annotations.go]: That's a good point, I want to maintain the same behavior as before, so if we changed it to `documentationInfo.description = ann.description` would do the trick?

**Comment by gcheadle-vmware** [comment]
> two different syntax for adding an example WITHOUT a description.
>
> @schema/examples (‘value’,)
>
> @schema/examples (“”, “value”)
> Since if the description is “” it will be ignored.

@cari-lynn 
I'm glad you pointed this out, I thought that it was not right to add an empty string as the value to the `x-example-description` key, so I treated the empty string as the default case. I could see a case where a GUI is displaying the example and it's description from an OpenAPI schema, and an author might need to always include the `x-example-description` key since the GUI checks for it. The author in that case may want the ability to include an empty string?

However the `x-example-description` key is [a custom extension to the OpenAPI spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#specification-extensions) so this GUI would have to be built specifically for the schema that `ytt` generates...

**Comment by pivotaljohn** [comment]
> @cari-lynn I'm glad you pointed this out, I thought that it was not right to add an empty string as the value to the `x-example-description` key, so I treated the empty string as the default case. I could see a case where a GUI is displaying the example and it's description from an OpenAPI schema, and an author might need to always include the `x-example-description` key since the GUI checks for it. The author in that case may want the ability to include an empty string?

Oh, as having consumed one-to-many APIs where keys disappear when no value is specified, I really appreciate this thinking. Reminds me of the Go Proverb, "[Make the zero value useful](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=385s)".

> However the `x-example-description` key is [a custom extension to the OpenAPI spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#specification-extensions) so this GUI would have to be built specifically for the schema that `ytt` generates...

And it's the nature of maps/dictionaries/key-value pairs that such entries can easily/safely be ignored if not needed. 👍🏻

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go]:
```suggestion
    = expected: tuple containing description and example
```
nit.

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go]: Do we have access to the type of the actual argument? We'll learn in the expectation that "tuple" is what we want. 

```suggestion
    = found: string for @schema/examples (at schema.yml:3)
```

nit: s/by/at/ ??

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go]:
```suggestion
    = expected: 2-tuple containing description (string) and example value (of expected type)
```

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go]:
```suggestion
    = found: 3-tuple argument in @schema/examples (by schema.yml:3)
    = expected: 2-tuple containing description (string) and example value (of expected type)
`
```

nit: pedantically, it's the requisite one argument, but has the wrong cardinality.

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go]:
```suggestion
    = found: float value for @schema/examples (at schema.yml:3)
    = expected: 2-tuple containing description (string) and example value (of expected type)
```

File: pkg/schema/annotations.go
Comment (pivotaljohn) [inline pkg/schema/annotations.go]: 🤔  Interesting. I had some expectation that there'd be a single slice of examples: a struct with description and value (i.e. mapping closely to what the user inputs)

Had you considered that approach and found it less favorable to storing the descriptions and examples separately?

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go]: I believe this would be a non-empty description _with_ a `None` value...?

I think you mean to hint: `(, "example value")`

File: pkg/schema/annotations.go
Comment (pivotaljohn) [inline pkg/schema/annotations.go]: If we negated the check for whether this is a tuple and pulled this up above the switch, it would pull in this whole body of code in by one indentation.

I think of checks like that (known also as "guard clauses") as a kind of pre-condition ... that once execution passes that point, the reader can safely assume the condition has been met: fewer things to keep track of in my head as I'm grokking the code.

File: pkg/schema/annotations.go
Comment (pivotaljohn) [inline pkg/schema/annotations.go]: 🤔 💭 
I wonder how this would read if all of the error conditions were captured in cases and the `default` is the success case...

```go
switch {
  case len(exampleTuple) == 0: // error
  case len(exampleTuple) == 1: // error
  case len(exampleTuple) > 2:  // error
  default: // success
}
```
And there's no rule that says that you _have_ to have a `default` case.... it might visually help if that code just follows the `switch` entirely?!?!

_(related to next comment...)_

File: pkg/schema/annotations.go
Comment (pivotaljohn) [inline pkg/schema/annotations.go:613]: Perhaps `CheckDocument()` needs a rename...?

ideas:
- `CheckNodes()` — hint that this is a tree?
- `CheckAST()` — explicitly call out the tree? (eeeehh... not sure how widely we use AST in places other than yamlmeta....)
- `CheckNode()` — is it implied that we mean both the Node and children?

File: pkg/schema/type.go
Comment (pivotaljohn) [inline pkg/schema/type.go]: This signature will make it not possible for code outside of `schema` to be able to set documentation...

... have you already considered providing setters for `Title` and `Description` and then maybe something like `AddExample(..)`? or `SetExamples(...)`?

Related:
- Today, we can only use the first example in OpenAPI v3.0 output
- But when we turn around and support OpenAPI v3.1 (and that _will_ happen), we'll want to be able to store multiple examples.
- Also, we'd love to be able to support _multiple_ examples in `ytt` Schema, today; no need to limit _that_ part of these system.
- Consider what happens when we push the limitation right up to the seam where it's introduced (i.e. when building the OpenAPI output).

**Review by pivotaljohn — CHANGES_REQUESTED** [review]

File: pkg/cmd/template/schema_author_test.go
Comment (cari-lynn) [reply pkg/cmd/template/schema_author_test.go]: (I looked into this a bit so im commenting what I found) I thought `('example value',)` was odd too. But I believe the trailing comma is the correct syntax. That makes it a single-element tuple. I believe the leading comma syntax isn't supported. 
 https://wiki.python.org/moin/TupleSyntax#:~:text=Multiple%20Element%20Tuples&text=Multiple%2Delement%20tuples%20may%20be,be%20enclosed%20in%20parentheses%2C%20e.g.

**Comment by cari-lynn** [comment]
> I thought that it was not right to add an empty string as the value to the x-example-description key, so I treated the empty string as the default case

I can see adding the `x-example-description` key adding noise to an already dense schema and not a lot of value. 

> And it's the nature of maps/dictionaries/key-value pairs that such entries can easily/safely be ignored if not needed. 👍🏻

Forgive me for being thick-skulled, do you mean that the empty string in the case of `("", "value")` can be ignored if not needed or do you mean that OpenAPI schema _consumers_ can ignore `x-example-description` if not needed?

For users that want to create a tuple without a description we could encourage users to use the `('example value',)` syntax, while having the `('', 'example value')` syntax behave as predicted and add `x-example-description: ""`.

**Comment by cari-lynn** [comment]
One more thing I was experimenting with is using a yamlFragment in the annotation to clean things up:

`#@schema/examples func()`

But I believe there is no way to specify a tuple in yaml.. which makes abstracting this to a function not possible using yamlFragements.

**Comment by pivotaljohn** [comment]
> > GC: I thought that it was not right to add an empty string as the value to the x-example-description key, so I treated the empty string as the default case
> 
> CLD: I can see adding the `x-example-description` key adding noise to an already dense schema and not a lot of value.

I agree that if its not typically used, it's more noise than signal. But by making it optional (i.e. sometimes present, sometimes not), makes that "API" a bit more complicated to use (noted, below).
 
> > JSR: And it's the nature of maps/dictionaries/key-value pairs that such entries can easily/safely be ignored if not needed. 👍🏻
> 
> CLD: Forgive me for being thick-skulled, do you mean that the empty string in the case of `("", "value")` can be ignored if not needed or do you mean that OpenAPI schema _consumers_ can ignore `x-example-description` if not needed?

The latter: that it's (usually) easy to ignore an entry in a map. I'm drawing an analogy with JSON APIs: that it makes for an easier interface to program to when every time there's an `example:` property, it _always_ has an `x-example-description`... even if an empty string. It's analogous to a field that's zero vs. not present/defined.

> For users that want to create a tuple without a description we could encourage users to use the `('example value',)` syntax, while having the `('', 'example value')` syntax behave as predicted and add `x-example-description: ""`.

Wouldn't that mean that _sometimes_ the first element in the tuple is the `example value` and other-times it would be the second element in the tuple?

File: pkg/schema/annotations.go
Comment (pivotaljohn) [reply pkg/schema/annotations.go:613]: handled in a subsequent PR. 👍🏻

**Comment by gcheadle-vmware** [comment]
After syncing with John, I decided to go with a simpler design for the annotation, where the description and example are always required. If the user does not want to include a description, they can just provide an empty string and the `x-example-description` key will just have an empty string.

File: pkg/cmd/template/schema_author_test.go
Comment (gcheadle-vmware) [reply pkg/cmd/template/schema_author_test.go]: See comment below about disallowing this for now.

**Comment by cari-lynn** [comment]
Commenting here to wrap up the threads I started. 
> [pivotalJohn] it always has an x-example-description... even if an empty string

Sounds good. I agree with the decisions above 👍  If we hear from users that this is noisy, in the future we can add a flag or syntax to omit this.

> [me] But I believe there is no way to specify a tuple in yaml.. which makes abstracting this to a function not possible using yamlFragements.

I am satisfied with being able to use a syntax to declare the `example` in a yamlFragment, and not necessarily being able to declare the entire tuple in a yamlFragment. Eg:
`#@schema/example ("host", hostExample())`

Thank you!!

**Comment by gcheadle-vmware** [comment]
@pivotaljohn, @cari-lynn - changes made to type interface to reflect the conversation earlier today.

File: pkg/cmd/template/schema_author_test.go
Comment (pivotaljohn) [inline pkg/cmd/template/schema_author_test.go:942]: Just note: the hint is misplaced and is fixed in #582. This is a new instance of the hint and will need to be removed after _that_ PR is merged.

**Review by pivotaljohn — APPROVED** [review]
Looks great, @gcheadle-vmware.

Note the minor interaction with #582 — there's a hint in here that (should) go away when _it_ merges.

Will leave it to you to decide which order you want these merged to minimize the effort on your part. Ping me if you want a buddy in that.

**Pull request merged**
